streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBWithTimestampStore.java

```diff
@@ -0,0 +1,679 @@
+            throw new ProcessorStateException(fatal);
+        }
+
+        open = true;
+    }
+
+    public void init(final ProcessorContext context,
+                     final StateStore root) {
+        // open the DB dir
+        internalProcessorContext = context;
+        openDB(context);
+        batchingStateRestoreCallback = new RocksDBBatchingRestoreCallback(this);
+
+        // value getter should always read directly from rocksDB
+        // since it is only for values that are already flushed
+        context.register(root, batchingStateRestoreCallback);
+    }
+
+    // visible for testing
+    boolean isPrepareForBulkload() {
+        return prepareForBulkload;
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public boolean persistent() {
+        return true;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return open;
+    }
+
+    @Override
+    public synchronized byte[] get(final Bytes key) {
+        validateStoreOpen();
+        return getInternal(key.get());
+    }
+
+    private void validateStoreOpen() {
+        if (!open) {
+            throw new InvalidStateStoreException("Store " + name + " is currently closed");
+        }
+    }
+
+    private byte[] getInternal(final byte[] rawKey) {
+        try {
+            final byte[] rawValueWithTimestamp = db.get(withTimestampColumnFamily, rawKey);
+            if (rawValueWithTimestamp != null) {
+                return rawValueWithTimestamp;
+            }
+
+            final byte[] rawValue = db.get(noTimestampColumnFamily, rawKey);
+            if (rawValue != null) {
+                final byte[] rawValueWithUnknownTimestamp = getValueWithUnknownTimestamp(rawValue);
+                // this does only work, because the changelog topic contains correct data already
+                // for other format changes, we cannot take this short cut and can only migrate data
+                // from old to new store on put()
+                putInternal(rawKey, rawValueWithUnknownTimestamp);
+                return rawValueWithUnknownTimestamp;
+            }
+
+            return null;
+        } catch (final RocksDBException e) {
+            // String format is happening in wrapping stores. So formatted message is thrown from wrapping stores.
+            throw new ProcessorStateException("Error while getting value for key %s from store " + name, e);
+        }
+    }
+
+    void toggleDbForBulkLoading(final boolean prepareForBulkload) {
+        if (prepareForBulkload) {
+            // if the store is not empty, we need to compact to get around the num.levels check for bulk loading
+            final String[] sstFileNames = dbDir.list((dir, name) -> SST_FILE_EXTENSION.matcher(name).matches());
+
+            if (sstFileNames != null && sstFileNames.length > 0) {
+                try {
+                    db.compactRange(true, 1, 0);
+                    db.compactRange(true, 1, 0);
```

Comment:
On oversight from my side. I _think_ we can just call it once (will remove the duplicate line).

The docs are not very specific if this triggers compaction for all CF, but I believe yes. There is also an API to compact one CF, but I don't see any reason to use it. Thoughts?

Reply:
Thinking about this one more, I actually believe that this would only compact default CF -- I'll add this to the `RocksDBAccessor` interface and pass in CF handlers.